examples/psp22-extension/runtime/psp22-extension-example.rs

```diff
@@ -0,0 +1,398 @@
+    log::{
+        error,
+        trace,
+    },
+    pallet_prelude::*,
+    traits::fungibles::{
+        approvals::{
+            Inspect as AllowanceInspect,
+            Mutate as AllowanceMutate,
+        },
+        Inspect,
+        InspectMetadata,
+        Transfer,
+    },
+};
+use pallet_assets::{
+    self,
+    WeightInfo,
+};
+use pallet_contracts::chain_extension::{
+    ChainExtension,
+    Environment,
+    Ext,
+    InitState,
+    RetVal,
+    SysConfig,
+    UncheckedFrom,
+};
+use sp_runtime::{
+    traits::{
+        Saturating,
+        StaticLookup,
+        Zero,
+    },
+    DispatchError,
+};
+
+#[derive(Debug, PartialEq, Encode, Decode, MaxEncodedLen)]
+struct Psp22BalanceOfInput<AssetId, AccountId> {
+    pub asset_id: AssetId,
+    pub owner: AccountId,
+}
+
+#[derive(Debug, PartialEq, Encode, Decode, MaxEncodedLen)]
+struct Psp22AllowanceInput<AssetId, AccountId> {
+    pub asset_id: AssetId,
+    pub owner: AccountId,
+    pub spender: AccountId,
+}
+
+#[derive(Debug, PartialEq, Encode, Decode, MaxEncodedLen)]
+struct Psp22TransferInput<AssetId, AccountId, Balance> {
+    pub asset_id: AssetId,
+    pub to: AccountId,
+    pub value: Balance,
+}
+
+#[derive(Debug, PartialEq, Encode, Decode, MaxEncodedLen)]
+struct Psp22TransferFromInput<AssetId, AccountId, Balance> {
+    pub asset_id: AssetId,
+    pub from: AccountId,
+    pub to: AccountId,
+    pub value: Balance,
+}
+
+#[derive(Debug, PartialEq, Encode, Decode, MaxEncodedLen)]
+struct Psp22ApproveInput<AssetId, AccountId, Balance> {
+    pub asset_id: AssetId,
+    pub spender: AccountId,
+    pub value: Balance,
+}
+
+pub struct Psp22Extension;
+
+fn map_err(err_msg: &'static str) -> impl FnOnce(DispatchError) -> DispatchError {
+    move |err| {
+        trace!(
+            target: "runtime",
+            "PSP22 Transfer failed:{:?}",
+            err
+        );
+        DispatchError::Other(err_msg)
+    }
+}
+
+fn metadata<T, E>(
+    func_id: u32,
+    env: Environment<E, InitState>,
+) -> Result<(), DispatchError>
+where
+    T: pallet_assets::Config + pallet_contracts::Config,
+    <T as SysConfig>::AccountId: UncheckedFrom<<T as SysConfig>::Hash> + AsRef<[u8]>,
+    E: Ext<T = T>,
+{
+    let mut env = env.buf_in_buf_out();
+    let asset_id = env.read_as()?;
+    let result = match func_id {
+        // PSP22Metadata::token_name
+        0x3d261bd4 => {
+            <pallet_assets::Pallet<T> as InspectMetadata<T::AccountId>>::name(&asset_id)
+                .encode()
+        }
+        // PSP22Metadata::token_symbol
+        0x34205be5 => {
+            <pallet_assets::Pallet<T> as InspectMetadata<T::AccountId>>::symbol(&asset_id)
+                .encode()
+        }
+        // PSP22Metadata::token_decimals
+        0x7271b782 => {
+            <pallet_assets::Pallet<T> as InspectMetadata<T::AccountId>>::decimals(
+                &asset_id,
+            )
+            .encode()
+        }
+        _ => unreachable!(),
```

Comment:
This isn't actually unreachable, someone can call `metadata` without doing any checks on the incoming `func_id`. I'd change this to an error instead

Reply:
Well, it's a private function of this module and it's called only once, and the call is made with a check. So I'm not sure what you have in mind when saying "someone can call `metadata` without doing any checks".

Reply:
OK, I've come up with a cleaner, enum-based solution to avoid this cases completely. LMKWYT